analyses/batch-effects/util/functions.R

```diff
@@ -0,0 +1,361 @@
+  covariate = covariate %>% select(1:4, batch, everything())
+  
+  return(covariate)
+}
+
+# read in the covariate data and create a tibble with two columns: the patient ID and the Batch
+make_id_batch = function(covariate, batch_column){
+  
+  # move the batch column to the front and only include RNA-Seq data
+  covariate = covariate_filtering(covariate, batch_column) 
+
+  # make a new matrix which will contain patient ID's and the corresponding batch
+  id_batch = cbind("Kids_First_Biospecimen_ID" = covariate$Kids_First_Biospecimen_ID, "batch" = covariate$batch)
+
+  # rename columns
+  colnames(id_batch) <- c("Kids_First_Biospecimen_ID", "batch")
+
+  # convert id_batch from a matrix to a tibble
+  id_batch = as_tibble(id_batch)
+  
+  return(id_batch)
+}
+
+make_id_batch_histology = function(covariate, batch_column){
+  # move the batch column to the front and only include RNA-Seq data
+  covariate = covariate_filtering(covariate, batch_column) 
+  
+  # make a new matrix which will contain patient ID's and the corresponding batch and histology
+  id_batch_histology = cbind("Kids_First_Biospecimen_ID" = covariate$Kids_First_Biospecimen_ID, "batch" = covariate$batch, "histology" = covariate$short_histology)
+  
+  #rename columns
+  colnames(id_batch_histology) <- c("Kids_First_Biospecimen_ID", "batch", "histology")
+  id_batch_histology = as_tibble(id_batch_histology)
+  
+  # convert id_batch_histology from a matrix to a tibble
+  id_batch_histology = as_tibble(id_batch_histology)
+  
+  return(id_batch_histology)
+}
+
+# make two PCA plots before and after applying combat
+make_before_and_after_combat_plots = function(data_t, batch, report_name, run_combat) {
+  
+  data_t$batch = batch_prep(data_t)
+  
+  #make a matrix representing just the gene expression data (this gets plugged into combat)
+  poly1matrix = t(data_t[,4:ncol(data_t)])
+  
+  pca = prcomp(log2(data_t[,4:ncol(data_t)]+ 1))
+  
+  tb = as_tibble(pca$x)
+  
+  p1 = make_plot(tb, data_t, report_name, "histology.png")
+  
+  if (run_combat == FALSE){
+    print("Combat set to not run on this data set")
+    return()
+  }
+  
+  plot_list = tryCatch({
+    # filter out genes without variance
+    var_zero_genes = poly1matrix[which((apply(poly1matrix, 1, var)==0)),]
+    poly1matrix = poly1matrix[which((apply(poly1matrix, 1, var)>0)),]
+    print(sum(var_zero_genes))
+    
+    poly2matrix <- log2(poly1matrix + 1)
+    
+    
+    combat_matrix = ComBat(dat=poly2matrix, batch=batch)
+    
+    pca = prcomp(t(combat_matrix))
+    
+    tb = as_tibble(pca$x)
+    
+    p2 = make_plot(tb, data_t, report_name, "batch-adjusted-histology.png")
+    return(list(p1, p2, data_t))
+  }, error=function(cond){
+    message("ComBat failed! It's likely that you are trying to adjust for batches with just 1 batch")
+    # message(cond)
+    return(list(p1, NULL, data_t))
+  })
+  
+  return(plot_list)
+  
+}
+
+# only include three most common histologies
```

Comment:
Is this because the three most common histologies happen to be represented in poly-A and stranded? This seems like a case where we might want to hardcode the `short_histology` values that are in both datasets.